### PR TITLE
Feature addonbar

### PR DESCRIPTION
**Pull request by AWolf81**
I think the new feature is almost ready to be merge to master but I have some open points I'd like to discuss.

(The addon bar adds the behviour as requested in issue #19 so you can easily test if your whitelist setting is working as expected.)

**First let me describe what the feature is doing at the moment:**
It adds a status icon in the addonbar (a icon in the upper right part of the browser) that will indicate if the links are enabled for the currently active tab. Green icon for enabled and red for disabled. (see Screenshots below).

A left click on the icon will open the addon preferences for easier access to the settings.
Switching between tabs will check if the page is in the whitelist of the addon and the icon will update accordingly.

(In the feature branch there is a drafts folder in data/img with my first version of the active/inactive icon that I'll delete later.)

screenshot active

screenshot inactive

For the following points I'd like to have a feedback:
- What do you think about the feature at all? Is it useful?
- Icon styling OK or should it be improved? Contrast isn't that good maybe should I make the colors a bit darker or any other ideas? Or maybe adding a darker border to the icon could help (same shape just darker color as border). Or pick a gray color like the other icons for inactive and green for active? Changed to gray inactive icon. (Edges still a bit blurry but I can't get them sharper. They're sharp in image editor. Maybe a resolution issue.)
- Should we add a preference to enable / disable the status icon?
- At tabs like `about:addons` or `about:newtabs` the icon will be red because the addon is not active which is correct. Is this behaviour acceptable?

### CODE REVIEW

**Comment by feinstaub** [comment]
> What do you think about the feature at all? Is it useful?

I think it is good to have some visual information about the addon.

> Icon styling OK or should it be improved?

If it possible I would suggest to reuse the addon's icon (doc/addon_icon_32x32.png). It's color is not as attention grabbing as the green from the screenshot. For "inactive" I would suggest grayscale. To get rid of blurry edges, it's maybe better to use svg graphics and/or already-created creative commons icons (e.g. see here https://icomoon.io/, I don't know if all icons are CC but there should be an information)

> Should we add a preference to enable / disable the status icon?

Not really needed because the user can use Firefox Menu -> Customize and remove the icon.

> At tabs like about:addons or about:newtabs the icon will be red because the addon is not active which is correct. Is this behaviour acceptable?

If the icon would be gray instead of red I see no problem here. No special handling if not necessary.

File: lib/utils/matchUrl.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * @param include object {0: "*", ...}
+ * @param uri to test
+ * @return testResult true or false
+ */
+exports.isUriIncluded = function(include, uri) {
+    // const specialUris = [
+    //     'about:newtab',
+    //     'about:addons'
+    // ];
+    var test = false;
```
Comment (feinstaub) [inline lib/utils/matchUrl.js]: could be renamed to "isIncluded"

File: lib/utils/matchUrl.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+var { MatchPattern } = require("sdk/util/match-pattern");
+
+/**
+ * Check if URI is in whitelist
+ * @param include object {0: "*", ...}
+ * @param uri to test
+ * @return testResult true or false
+ */
+exports.isUriIncluded = function(include, uri) {
```
Comment (feinstaub) [inline lib/utils/matchUrl.js]: instead of "include" you can write "includePatternMap" (or ...Dict or ...List)

File: lib/utils/matchUrl.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    // else {
+    Object.keys(include).forEach(function (key) {
+        var pattern = new MatchPattern(include[key]);
+
+        // use val
+        // console.log('pattern = ' + include[key]);
+        if ( pattern.test(uri) ) {
+            test = true;
+            // console.log('pattern matched ', include[key]);
+            return test; // exit at first match and return true
```
Comment (feinstaub) [inline lib/utils/matchUrl.js]: you could remove the test variable completely. Just return true.

File: lib/utils/matchUrl.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    });
+    //}
+
+    return test; // if we're here --> no pattern matched
```
Comment (feinstaub) [inline lib/utils/matchUrl.js]: just return false

File: lib/utils/matchUrl.js (hunk 2 above)
Comment (AWolf81) [inline lib/utils/matchUrl.js]: I think `includePattern` is also OK?

File: lib/utils/matchUrl.js (hunk 3 above)
Comment (AWolf81) [inline lib/utils/matchUrl.js]: Thanks. You're right. Removed `test` variable and changed the returns.

File: lib/utils/matchUrl.js (hunk 2 above)
Comment (feinstaub) [inline lib/utils/matchUrl.js]: The variable name should convey that it is a list of some sort (which it is, isn't it?). See e.g. here for discussion about naming: http://stackoverflow.com/questions/395739/do-you-name-your-arrays-plurally-or-singularly, They would name it includePatterns or includePatternList,

Comment (AWolf81) [inline lib/utils/matchUrl.js]: OK, I've renamed it to includePatterns. (It is an array-like object of include URIs)

**Comment by AWolf81** [comment]
I think the icon is OK now and it's good to reuse the icon here.  What do you think? See updated screenshots above.

That's right. The removal of the icon is easy just modify the addonbar and move the icon away from the addonbar.
During that test I've found one small bug. If I remove the icon from the addonbar and restart firefox the updating of the addon status isn't working anymore. Need to check how to fix this.

File: lib/utils/matchUrl.js [hunk 5]
```diff
@@ -20,14 +20,14 @@ exports.isUriIncluded = function(includePattern, uri) {
...
+        // console.log(includePatterns[key]);
+        // @todo includePatterns[key] can throw an error if user passes a whitespace
         //       in front of *.url and it fails silently at the moment.
-        var pattern = new MatchPattern(includePattern[key]);
+        var pattern = new MatchPattern(includePatterns[key]);
 
         // use val
-        // console.log('pattern = ' + includePattern[key]);
+        // console.log('pattern = ' + includePatterns[key]);
         return ( pattern.test(uri) );
```
Comment (feinstaub) [inline lib/utils/matchUrl.js]: nit-pick: no spaces after ( and before ). In this case, they are not needed at all.

**Comment by feinstaub** [comment]
> I think the icon is OK now and it's good to reuse the icon here. What do you think? See updated screenshots above.

Thanks for updating. It looks really good, i.e. consistent. (One might say that it is the only colored icon in the toolbar, but I think we can leave it for now. Other addons like Adblock Plus have not updated to the new monochromatic theme, too)

> During that test I've found one small bug. [...]

Rest is fine with me. I'll start preparing the new version as soon as you decide to merge the PR.

**Comment by AWolf81** [comment]
The mentioned bug isn't a bug. It was a mistake from me because starting  a new tab will show the icon inactive and that's correct because there is no url in the address bar and I've expected the icon to be active.

So the icon is also updating correctly even it's not in the addonbar.

**Pull request merged**